src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam.BamErrorMetric;
+
+import htsjdk.samtools.reference.SamLocusAndReferenceIterator.SAMLocusAndReference;
+import htsjdk.samtools.util.CollectionUtil;
+import htsjdk.samtools.util.SamLocusIterator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeSet;
+import java.util.function.Supplier;
+
+/**
+ * An interface and implementations for classes that apply a {@link picard.sam.BamErrorMetric.ReadBaseStratification.RecordAndOffsetStratifier RecordAndOffsetStratifier}
+ * to put bases into various "bins" and then compute an {@link ErrorMetrics} on these bases using a {@link BaseErrorCalculation}.
+ *
+ * @author Yossi Farjoun
+ */
+
+public class BaseErrorAggregation<CALCULATOR extends BaseErrorCalculation.BaseCalculator> {
+    private final Supplier<CALCULATOR> simpleAggregatorGenerator;
+    private final ReadBaseStratification.RecordAndOffsetStratifier stratifier;
+
+    // Updated as a defaulting map.
+    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+    private final CollectionUtil.DefaultingMap<Object, CALCULATOR> strataAggregatorMap = new CollectionUtil.DefaultingMap<>
+            (new CollectionUtil.DefaultingMap.Factory<CALCULATOR, Object>() {
+                @Override
+                public CALCULATOR make(final Object ignored) {
+                    return simpleAggregatorGenerator.get();
+                }
+            }, true);
+
+    public BaseErrorAggregation(final Supplier<CALCULATOR> simpleAggregatorGenerator,
+                                final ReadBaseStratification.RecordAndOffsetStratifier stratifier) {
+        this.stratifier = stratifier;
+        this.simpleAggregatorGenerator = simpleAggregatorGenerator;
+    }
+
+    public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SAMLocusAndReference locusInfo) {
+        final Object stratus = stratifier.stratify(recordAndOffset, locusInfo);
+        // this assumes we do not want to aggregate null.
+        if (stratus != null) strataAggregatorMap.get(stratus).addBase(recordAndOffset, locusInfo);
+
+    }
+
+    public String getSuffix() {
+        return simpleAggregatorGenerator.get().getSuffix() + "_by_" + stratifier.getSuffix();
+    }
+
+    public ErrorMetrics.ErrorMetricBase[] getMetrics() {
+        final List<ErrorMetrics.ErrorMetricBase> metrics = new ArrayList<>();
+
+        // we do this to sort
+        final TreeSet<Object> strata = new TreeSet<>(strataAggregatorMap.keySet());
```

Comment:
This is N^2, if you instead create a list and then sort it it would be more efficient. Not sure if that matters here though.

Reply:
You're right, it's NlogN. I thought add() was N, but it makes sense that it's logN